### PR TITLE
Tiny STV parsing tests all passing.

### LINKED ISSUE #107: Drop STV CVRs
Although we are not required to audit STV contests (i.e. multi-winner preferential contests), we need to ensure that colorado-rla does not break when they are uploaded. This requires a bit of thought. Probably we want three types of contests: plurality, IRV and STV.
Probably:
- when parsing, a contest with ranks and multiple winners should be STV,
- STV contests should be always NOT_AUDITABLE
- STV contests cannot be targeted for audit
- STV winners should have a clear placeholder (e.g. UNKNOWN or "") - note that the default contest counter will get the wrong winner, just as it does for IRV.
- STV estimatedSampleSizes and optimisticSampleSizes should match the values for other NOT_AUDITABLE contests. (I have a vague recollection that this is 0, but we should check.)

Think about whether we could easily get a count of 'other discrepancy' in the event that an STV contest happens to be on an audited ballot.

Consider whether all this is best achieved with an STVComparisonAudit class.

### PR DESCRIPTION
**Pull request by vteague**
Drops all trace of STV contests, as if they'd been cut out of the CVR before upload.

### CODE REVIEW

**Review by cjcarlton — APPROVED** [review]
Changes and tests LGTM! I briefly considered if the booleans (isPlurality, isIRV and isSTV) could be simplified a bit (perhaps using !isPlurality in the definitions for isIRV and isSTV but it's possible that would miss a 4th possibility of all 3 booleans being false.

File: server/eclipse-project/src/main/java/us/freeandfair/corla/csv/DominionCVRExportParser.java [hunk 1]
```diff
@@ -334,19 +351,26 @@ private void updateContestStructures(final CSVRecord the_line,
...
-          String contestName = c.substring(0, c.indexOf(PLURALITY_VOTE_FOR)).strip();
+        final boolean isPlurality = pluralityVotesAllowed > 0 && irvVotesAllowed == -1 && irvWinners == -1;
+        final boolean isIRV = pluralityVotesAllowed == -1 && irvVotesAllowed > 0 && irvWinners == 1;
+        // If it looks like IRV but has more than one winner, it's an STV contest.
+        final boolean isSTV = pluralityVotesAllowed == -1 && irvVotesAllowed > 0 && irvWinners > 1;
+
+        // Counterintuitively, we also code the STV contest as a 'plurality' contest, which works
+        // fine with candidate names of the form Alice(1), Alice(2), etc.
+        // STV_NO_VOTES = -1 in the_votes_allowed to encode that it's an STV contest.
+        if(isPlurality || isSTV) {
```
Comment (michelleblom) [inline server/eclipse-project/src/main/java/us/freeandfair/corla/csv/DominionCVRExportParser.java:363]: I think contestName can be final.

File: server/eclipse-project/src/main/java/us/freeandfair/corla/csv/DominionCVRExportParser.java [hunk 2]
```diff
@@ -334,19 +351,26 @@ private void updateContestStructures(final CSVRecord the_line,
...
-          the_votes_allowed.put(contestName, pluralityVotesAllowed);
+          the_votes_allowed.put(contestName, isPlurality ? pluralityVotesAllowed : STV_NO_VOTES);
 
         // If winners and allowed votes are as expected for IRV, this is an IRV contest.
-        } else if(pluralityVotesAllowed == -1 && irvVotesAllowed > 0 && irvWinners == 1
-            // We expect the count to be the real number of choices times the number of ranks.
-            && count % irvVotesAllowed == 0) {
+        // We expect the count to be the real number of choices times the number of ranks.
+        } else if(isIRV && count % irvVotesAllowed == 0) {
           String contestName = c.substring(0, c.indexOf(IRV_WINNERS_ALLOWED)).strip();
```
Comment (michelleblom) [inline server/eclipse-project/src/main/java/us/freeandfair/corla/csv/DominionCVRExportParser.java]: contestName could be final.

**Comment by vteague** [comment]
> Changes and tests LGTM! I briefly considered if the booleans (isPlurality, isIRV and isSTV) could be simplified a bit (perhaps using !isPlurality in the definitions for isIRV and isSTV but it's possible that would miss a 4th possibility of all 3 booleans being false.

Yes, indeed, those three cases do not cover all possible values @charliecarlton, only the ones we expect to encounter. I guess I decided that, on something a bit unexpected, it's probably better to raise an error than to let it past. The obvious example would be something that has both a "Vote For=" and a "Number of positions=" header. This would actually make perfect sense as a plurality contest for which the number of available positions doesn't match the number you're permitted to vote for. There's no particular reason they need to be the same - they just are the same at the moment for all the contests in CO. 

So the code I've written will throw an error at that particular combination of options. I _think_ this is the right thing to do, because if we let it through I'm not sure that the winners_allowed and votes_allowed options would be set properly (they probably wouldn't). In short, if some new combination of options appears, I think somebody probably needs to go through and make sure it does all the right things.

So there's a much simpler way of writing it, like
```
        // If winners and allowed votes are as expected for plurality, this is a plurality contest.
        final boolean isPlurality = pluralityVotesAllowed > 0;
        final boolean isIRV = !isPlurality && irvVotesAllowed > 0 && irvWinners == 1;
        // If it looks like IRV but has more than one winner, it's an STV contest.
        final boolean isSTV = !isPlurality && irvVotesAllowed > 0 && irvWinners > 1;
```
But this lacks the future-defensiveness as above. I can't think of a nicer way of writing it that is both easier to read _and_ defensive against future unexpected combinations.

File: server/eclipse-project/src/test/java/au/org/democracydevelopers/corla/csv/DominionCVRExportParserTests.java [hunk 3]
```diff
@@ -611,4 +619,167 @@ public void parseWriteIns() throws IOException {
...
+   */
+  @Test
+  @Transactional
+  public void parseSTVPlusThreeCandidatesTenVotesPlusPluralitySucceedsAndDropsSTV() throws IOException {
+    testUtils.log(LOGGER, "parseSTVPlusThreeCandidatesTenVotesPlusPluralitySucceedsAndDropsSTV");
+    final Path path = Paths.get(TINY_CSV_PATH + "STVPlusThreeCandidatesTenVotesPlusPlurality.csv");
+
+    doIRVAndSTVAndPluralityTest(path, "La Plata");
+  }
+
```
Comment (michelleblom) [inline server/eclipse-project/src/test/java/au/org/democracydevelopers/corla/csv/DominionCVRExportParserTests.java:665]: Rogue extra line at top of method comment for noContestUploadIsAnError(). Also, some remaining tests have no public, just void.

File: server/eclipse-project/src/test/java/au/org/democracydevelopers/corla/csv/DominionCVRExportParserTests.java [hunk 4]
```diff
@@ -611,4 +619,167 @@ public void parseWriteIns() throws IOException {
...
+  void allSTVUploadIsOK() throws IOException {
+    testUtils.log(LOGGER, "allSTVUploadIsOK");
+    final Path path = Paths.get(TINY_CSV_PATH + "TinySTVOnly.csv");
+    final Reader reader = Files.newBufferedReader(path);
+
+    County bent = fromString("Bent");
+    final DominionCVRExportParser parser = new DominionCVRExportParser(reader, bent, blank, true);
+    assertTrue(parser.parse().success);
+  }
+
```
Comment (michelleblom) [inline server/eclipse-project/src/test/java/au/org/democracydevelopers/corla/csv/DominionCVRExportParserTests.java:701]: The doIRVandSTVPluralityTest() method can be private.

**Review by michelleblom — APPROVED** [review]

**Pull request merged**